testplan/common/utils/table.py

```diff
@@ -1,40 +1,114 @@
+# pylint: enable=import-error, no-name-in-module
+
+_TP_BLANK_CELL = "_TP_BLANK_CELL"
+
+
+def _is_nonstr_sequence(obj):
+    return isinstance(obj, Sequence) and not isinstance(obj, six.string_types)
+
+
+def _table_from_list_of_lists(list_of_lists):
+    # type: (Sequence[Sequence[Any]]) -> List[List[Any]]
+    std_tbl = []
+    # all that needs to happen here is that each nested list must be padded to
+    # the length of the longest list
+    tbl_width = 0
+    # this bit finds the longest row and filters out invalid elements
+    for row in list_of_lists:
+        # TODO: determine whether we should raise here, or whether we should
+        #       try harder to guess what is meant by a non-list-like element
+        if not _is_nonstr_sequence(row):
+            continue
+        new_row = list(row)
+        std_tbl.append(new_row)
+        tbl_width = max(tbl_width, len(new_row))
+    for row in std_tbl:
+        pad_cells = tbl_width - len(row)
+        row.extend([_TP_BLANK_CELL] * pad_cells)
+    if len(std_tbl) == 0:  # empty list
+        return [[]]
+    return std_tbl
+
+
+def _table_from_list_of_dicts(list_of_dicts):
+    dict_of_lists = OrderedDict()
+    for dict_elem in list_of_dicts:
+        for hdr, val in six.iteritems(dict_elem):
+            dict_of_lists.setdefault(hdr, []).append(val)
+    return _table_from_dict_of_lists(dict_of_lists)
+
+
+def _table_from_dict_of_lists(dict_of_lists):
+    # type: (Mapping[AnyStr, Any]) -> List[List[Any]]
+    tbl_headers = list(six.iterkeys(dict_of_lists))
+    n_hdrs = len(tbl_headers)
+    if n_hdrs == 0:  # empty dict
+        return [[]]
+    new_prefilled_row = lambda: [_TP_BLANK_CELL] * n_hdrs
+    tbl_data = [new_prefilled_row()]
+    for col_idx, col_name in enumerate(tbl_headers):
+        val = dict_of_lists[col_name]
+        col_cells = list(val) if _is_nonstr_sequence(val) else [val]
+        for row_idx, cell_val in enumerate(col_cells):
+            # 1st iteration:
+            #   (a) `row_idx` == 0, `len(tbl_data)` == 1
+            #   (b) set `tbl_data[0][col_idx]` to `cell_val`
+            # 2nd iteration:
+            #   (a) `row_idx` == 1, `len(tbl_data)` == 1
+            #   (b) inject a filler row into `tbl_data`
+            #   (c) `row_idx` == 1, `len(tbl_data)` == 2 now
+            #   (d) set `tbl_data[1][col_idx]` to `cell_val`
+            # and so on..
+            if len(tbl_data) == row_idx:
+                tbl_data.append(new_prefilled_row())
+            tbl_data[row_idx][col_idx] = cell_val
+    return [tbl_headers] + tbl_data
 
 
 class TableEntry(object):
     """
-    Represents a table. Internally represented either
-    as a ``list`` of ``list`` or a ``list`` of ``dict``.
+    Represents a table. Internally represented as ``List[List[Any]]``.
     """
 
     def __init__(self, table):
-        self.table = self._validate_table(table or [])
-
-    @staticmethod
-    def _validate_table(table):
```

Comment:
Actually this original implementation that validates user input looks just fine to me. Why are we changing it?

With that said, one thing I may want to add is - if user passes a list of list, we may want all sub-list to be of uniform length.